zipkintracing/tracing.go

```diff
@@ -0,0 +1,217 @@
+package zipkintracing
+
+import (
+	"fmt"
+	"github.com/labstack/echo/v4/middleware"
+	"net/http"
+	"strconv"
+
+	"github.com/labstack/echo/v4"
+	"github.com/openzipkin/zipkin-go"
+	zipkinhttp "github.com/openzipkin/zipkin-go/middleware/http"
+	"github.com/openzipkin/zipkin-go/model"
+	"github.com/openzipkin/zipkin-go/propagation/b3"
+	"github.com/openzipkin/zipkin-go/reporter"
+	zipkinHttpReporter "github.com/openzipkin/zipkin-go/reporter/http"
+)
+
+type (
+
+	// tracer struct
+	tracer struct {
+		URL         string
+		ServiceName string
+		tracer      *zipkin.Tracer
+		client      *zipkinhttp.Client
+		reporter    reporter.Reporter
+	}
+
+	// Tracer interface tracing
+	Tracer interface {
+		init() error
+		GetTraceServerFunc(tags map[string]string) echo.MiddlewareFunc
+		GetTraceServerFuncWithConfig(tags map[string]string, config TraceServerFuncConfig) echo.MiddlewareFunc
+		GetTraceProxyFunc(tags map[string]string) echo.MiddlewareFunc
+		GetTraceProxyFuncWithConfig(tags map[string]string, config TraceProxyFuncConfig) echo.MiddlewareFunc
+		TraceFunc(c echo.Context, spanName string, tags map[string]string) func()
+		StartChildSpan(c echo.Context, spanName string, tags map[string]string) (childSpan zipkin.Span)
+		Close() error
+		HTTPDoer
+	}
+
+	//HTTPDoer is a wrapper interface for http
+	HTTPDoer interface {
+		DoHTTP(c echo.Context, r *http.Request) (*http.Response, error)
+	}
+
+	TraceServerFuncConfig struct {
+		Skipper middleware.Skipper
+	}
+
+	TraceProxyFuncConfig struct {
+		Skipper middleware.Skipper
+	}
+)
+
+var (
+	DefaultTraceServerFuncConfig = TraceServerFuncConfig{
+		Skipper: middleware.DefaultSkipper,
+	}
+
+	DefaultTraceProxyFuncConfig = TraceProxyFuncConfig{
+		Skipper: middleware.DefaultSkipper,
+	}
+)
+
+// init function to initialize zipkin tracer
+func (t *tracer) init() error {
+	t.reporter = zipkinHttpReporter.NewReporter(t.URL)
+
+	// create our local service endpoint
+	endpoint, err := zipkin.NewEndpoint(t.ServiceName, "")
+	if err != nil {
+		return err
+	}
+	// initialize our tracer
+	tracer, err := zipkin.NewTracer(t.reporter, zipkin.WithLocalEndpoint(endpoint))
+	if err != nil {
+		return err
+	}
+	client, err := zipkinhttp.NewClient(tracer, zipkinhttp.ClientTrace(false))
+	if err != nil {
+		return err
+	}
+	t.client = client
+	t.tracer = tracer
+	return nil
+}
+
+// NewTracer returns instance of tracer with client initialized
+func NewTracer(zipkinURL, serviceName string) (Tracer, error) {
+	tracer := &tracer{
+		URL:         zipkinURL,
+		ServiceName: serviceName,
+	}
+	err := tracer.init()
+	if err != nil {
+		return nil, err
+	}
+	return tracer, nil
+}
+
+// DoHTTP is a http zipkin tracer implementation of HTTPDoer
+func (t *tracer) DoHTTP(c echo.Context, r *http.Request) (*http.Response, error) {
+	req := r.WithContext(c.Request().Context())
+	return t.client.DoWithAppSpan(req, fmt.Sprintf("%s %s", r.Method, r.URL.Path))
```

Comment:
I'd rather not use r.URL.Path as part of the span name a it is high cardinality. In a HTTP client operation we should just use the verb.

Reply:
fixed and also added a DoHTTP test.